plotly/tools.py

```diff
@@ -1269,3 +1277,628 @@ def return_figure_from_figure_or_data(figure_or_data, validate_figure):
+                    else:
+                        break
+                if stotal > 2:
+                    break
+            return stotal, xf_traj, yf_traj
+
+        sf, xf_traj, yf_traj = rk4(x0, y0, f)
+        sb, xb_traj, yb_traj = rk4(x0, y0, g)
+        stotal = sf + sb
+        x_traj = xb_traj[::-1] + xf_traj[1:]
+        y_traj = yb_traj[::-1] + yf_traj[1:]
+
+        if len(x_traj) < 1:
+            return None
+        if stotal > .2:
+            initxb, inityb = self.blank_pos(x0, y0)
+            self.blank[inityb, initxb] = 1
+            return x_traj, y_traj
+        else:
+            for xb, yb in zip(xb_changes, yb_changes):
+                self.blank[yb, xb] = 0
+            return None
+
+    def traj(self, xb, yb):
+        """
+        Integrate trajectories, used in get_streamlines()
+        """
+        if xb < 0 or xb >= self.density or yb < 0 or yb >= self.density:
+            return
+        if self.blank[yb, xb] == 0:
+            t = self.rk4_integrate(xb*self.spacing_x, yb*self.spacing_y)
+            if t is not None:
+                self.trajectories.append(t)
+
+    def get_streamlines(self):
+        """
+        Get streamlines by building trajectory set.
+
+        :rtype (list of lists) st_x: lists of x-values for each streamline
+        :rtype (list of lists) st_y: lists of y-values for each streamline
+        """
+        for indent in range((max(self.density, self.density))//2):
+            for xi in range(max(self.density, self.density)-2*indent):
+                self.traj(xi+indent, indent)
+                self.traj(xi+indent, self.density-1-indent)
+                self.traj(indent, xi+indent)
+                self.traj(self.density-1-indent, xi+indent)
+
+        self.st_x = [np.array(t[0])*self.delta_x+self.x[0] for t in
+                     self.trajectories]
+        self.st_y = [np.array(t[1])*self.delta_y+self.y[0] for t in
+                     self.trajectories]
+
+        for index in range(len(self.st_x)):
+            self.st_x[index] = self.st_x[index].tolist()
+            self.st_x[index].append(np.nan)
+
+        for index in range(len(self.st_y)):
+            self.st_y[index] = self.st_y[index].tolist()
+            self.st_y[index].append(np.nan)
+
+    def get_streamline_arrows(self):
+        """
+        Makes an arrow for each streamline. Gets angle of streamline at 1/3
+        mark and creates arrow coordinates based off of user defined angle and
+        arrow_scale
+
+        :param (array) st_x: x-values for all streamlines
+        :param (array) st_y: y-values for all streamlines
+        :param (angle in radians) angle: angle of arrowhead. Default = pi/9
+        :param (float in [0,1]) arrow_scale: value to scale length of arrowhead
+            Default = .09
+        :rtype (list) arrows_x: x-values to create arrowhead
+        :rtype (list) arrows_y: y-values to create arrowhead
+        """
+        ArrowEnd_x = np.empty((len(self.st_x)))
```

Comment:
pep:8ball: naming conventions ~~ArrowEnd_x~~ --> `arrow_end_x`

Caps are typically reserved for Classes (like `TraceFactory`)